src/node_zip.cc

```diff
@@ -0,0 +1,454 @@
+  ZipArchive* zip;
+  zip_uint32_t index;
+  ZipArchive::CheckZipEntry(args, zip, index);
+
+  CHECK_EQ(zip_delete(zip->zip_, index), 0);
+}
+
+void ZipArchive::ReadEntry(const v8::FunctionCallbackInfo<v8::Value>& args) {
+  v8::Isolate* isolate = args.GetIsolate();
+
+  ZipArchive* zip;
+  zip_uint32_t index;
+  ZipArchive::CheckZipEntry(args, zip, index);
+
+  zip_stat_t file_stat;
+  CHECK_EQ(zip_stat_index(zip->zip_, index, 0, &file_stat), 0);
+  CHECK_LE(file_stat.comp_size, ZIP_INT64_MAX);
+  CHECK_LE(file_stat.comp_size, std::numeric_limits<size_t>::max());
+
+  char* data = reinterpret_cast<char*>(malloc(file_stat.comp_size * sizeof(char)));
+
+  zip_file_t* file;
+  CHECK_NE(file = zip_fopen_index(zip->zip_, index, ZIP_FL_COMPRESSED), nullptr);
+
+  zip_int64_t read_length = zip_fread(file, data, ZIP_INT64_MAX);
+  CHECK_EQ(read_length, static_cast<zip_int64_t>(file_stat.comp_size));
+
+  v8::Local<v8::Value> result[] = {
+    v8::Integer::NewFromUnsigned(isolate, file_stat.comp_method),
+    Buffer::New(isolate, data, read_length).ToLocalChecked(),
+  };
+
+  args.GetReturnValue().Set(v8::Array::New(isolate, &result[0], sizeof(result) / sizeof(result[0])));
+}
+
+void ZipArchive::StatEntry(const v8::FunctionCallbackInfo<v8::Value>& args) {
+  v8::Isolate* isolate = args.GetIsolate();
+
+  ZipArchive* zip;
+  zip_uint32_t index;
+  ZipArchive::CheckZipEntry(args, zip, index);
+
+  zip_stat_t file_stat;
+  CHECK_EQ(zip_stat_index(zip->zip_, index, 0, &file_stat), 0);
+  CHECK_LE(file_stat.comp_size, ZIP_INT64_MAX);
+  CHECK_LE(file_stat.comp_size, std::numeric_limits<size_t>::max());
+
+  zip_uint8_t opsys;
+  zip_uint32_t attributes;
+  CHECK_EQ(zip_file_get_external_attributes(zip->zip_, index, 0, &opsys, &attributes), 0);
+
+  v8::Local<v8::Value> result[] = {
+    v8::Integer::NewFromUnsigned(isolate, file_stat.size),
+    v8::Integer::NewFromUnsigned(isolate, file_stat.comp_size),
+    v8::Integer::NewFromUnsigned(isolate, file_stat.comp_method),
+    v8::Integer::NewFromUnsigned(isolate, file_stat.mtime),
+    v8::Integer::NewFromUnsigned(isolate, file_stat.crc),
+    v8::Integer::NewFromUnsigned(isolate, opsys),
+    v8::Integer::NewFromUnsigned(isolate, attributes),
+  };
+
+  args.GetReturnValue().Set(v8::Array::New(isolate, &result[0], sizeof(result) / sizeof(result[0])));
+}
+
+void ZipArchive::RestatEntry(const v8::FunctionCallbackInfo<v8::Value>& args) {
+  v8::Isolate* isolate = args.GetIsolate();
+  Environment* env = Environment::GetCurrent(isolate);
+  v8::Local<v8::Context> ctx = env->context();
+
+  ZipArchive* zip;
+  zip_uint32_t index;
+  ZipArchive::CheckZipEntry(args, zip, index);
+
+  const int argc = args.Length();
+  CHECK_GE(argc, 2);
+
+  CHECK(args[1]->IsObject());
+  v8::Local<v8::Object> stats = args[1]->ToObject(ctx).ToLocalChecked();
+
+  v8::Local<v8::Value> mtime;
+  if (stats->Get(ctx, v8::String::NewFromUtf8Literal(isolate, "mtime")).ToLocal(&mtime)) {
```

Comment:
You [would want to `return;` here](https://github.com/nodejs/node/blob/main/src/README.md#exception-handling) if `.ToLocal()` returns `false`, not do another `stats->Get` call

Reply:
Ah thanks! I wasn't sure whether a property being missing would return an empty Maybe, but I see now it's only when exceptions occur.